### PR TITLE
Policy plugin extn 

### PR DESCRIPTION
**Pull request by rajansandeep**
A very basic idea on how the "permit" and "deny" extn to the policy plugin will work.
Doesn't completely work.

@fturib

### CODE REVIEW

File: policy.go [hunk 1]
```diff
@@ -659,13 +741,32 @@ func (p *policyPlugin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dn
...
-		status = dns.RcodeServerFailure
-		goto Exit
-	}
+	if p.rule.keyword != "" {
+		result, err := p.evaluateExp(p.rule.params, p.rule.keyword, w, r)
+		if err != nil {
+			resolveFailed = true
+			goto Exit
+		}
+		if result == true {
```
Comment (fturib) [inline policy.go]: p.evaluateExp(..) should be pushed the PolycRule

=> p.policyRule[x].evaluate reply directly the action (typeAllow, typeRefuse or other ...)

File: policy.go [hunk 2]
```diff
@@ -659,13 +741,32 @@ func (p *policyPlugin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dn
...
+				ah.action = typeRefuse
+			}
+		}
+		if result == false {
+			ah.action = typeRefuse
+		}
+	}
+	if p.pdpSvr.policyFile != "" {
+		// validate domain name (validation #1)
+		err = p.validate(ah, "")
```
Comment (fturib) [inline policy.go]: This part should be a specific PolicyRule

File: policy.go [hunk 3]
```diff
@@ -86,6 +94,7 @@ type pdpServer struct {
 // requests and replies using PDP server.
 type policyPlugin struct {
 	endpoints       []string
+	rule            policyRule
```
Comment (fturib) [inline policy.go]: we expect : rules []PolicyRule
it is an array and the mechanism is to loop in order over this array until having a firm reply.

File: policy.go [hunk 4]
```diff
@@ -216,6 +228,76 @@ func (p *policyPlugin) parsePDP(c *caddy.Controller) error {
...
+
+func (p *policyPlugin) evaluateExpression(args []string) (interface{}, error) {
+	expression, err := govaluate.NewEvaluableExpression(strings.Join(args, " "))
+	if err != nil {
+		return nil, err
+	}
+	parameters := make(map[string]interface{}, 8)
+	parameters["queryName"] = args[0]
+	result, err := expression.Evaluate(parameters)
+	if err != nil {
```
Comment (fturib) [inline policy.go:257]: yes with several points:

1- parameters will be extracted from context (and you need Metadata)
2- you need to verify you have a boolean as a result. Or ensure to make the result a boolean

**Review by fturib — COMMENTED** [review]
I provided immediate comments.
I propose we have a session together this afternoon.

File: policy.go [hunk 5]
```diff
@@ -659,13 +741,32 @@ func (p *policyPlugin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dn
...
+				ah.action = typeAllow
 
+			} else if p.rule.keyword == "deny" {
+				ah.action = typeRefuse
+			}
+		}
+		if result == false {
+			ah.action = typeRefuse
+		}
+	}
```
Comment (yuewko) [inline policy.go]: The above block L744-L769 should be a method of the policyRule object, which returns the result of the evaluation.

File: policy.go [hunk 6]
```diff
@@ -216,6 +228,76 @@ func (p *policyPlugin) parsePDP(c *caddy.Controller) error {
...
+func (p *policyPlugin) parsePermit(c *caddy.Controller) error {
+	args := c.RemainingArgs()
+	argsLen := len(args)
+	if argsLen < 0 {
+		return c.ArgErr()
+	}
+	p.rule.params = args
+	p.rule.keyword = "permit"
+
+	_, err := p.evaluateExpression(args)
```
Comment (yuewko) [inline policy.go]: Not sure you want to call `evaluateExpression` here. At this point, you don't have any of the parameters that is obtained from the query.

File: policy.go [hunk 7]
```diff
@@ -216,6 +228,76 @@ func (p *policyPlugin) parsePDP(c *caddy.Controller) error {
...
+	}
+	p.rule.params = args
+	p.rule.keyword = "deny"
+
+	_, err := p.evaluateExpression(args)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (yuewko) [inline policy.go:319]: `parseDeny()` and `parsePermit()` should be consolidated into one method.

**Pull request closed without merging**